lib/Conversion/LinalgToArith/LinalgToArith.cpp

```diff
@@ -0,0 +1,385 @@
+#include "mlir/include/mlir/Transforms/DialectConversion.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
+
+#define DEBUG_TYPE "linalg-to-arith"
+
+namespace mlir {
+namespace heir {
+namespace linalg {
+
+#define GEN_PASS_DEF_LINALGTOARITH
+#include "lib/Conversion/LinalgToArith/LinalgToArith.h.inc"
+
+bool isPowerOfTwo(int64_t n) { return (n > 0) && ((n & (n - 1)) == 0); }
+
+int calculateIndexHelper(bool isLeftOperandSecret, int dim0, int dim1, int i,
+                         int j) {
+  // Note that we are producing the transpose of the diagonalized matrix.
+  if (isLeftOperandSecret) {
+    return ((i + j) % dim0) * dim1 + (j % dim1);
+  } else {  // right operand is secret
+    return (i % dim0) * dim0 + ((i + j) % dim1);
+  }
+}
+
+struct ConvertLinalgMatmul : public OpRewritePattern<mlir::linalg::MatmulOp> {
+ private:
+  DataFlowSolver *solver;
+
+ public:
+  ConvertLinalgMatmul(DataFlowSolver *solver, mlir::MLIRContext *context)
+      : OpRewritePattern<mlir::linalg::MatmulOp>(context), solver(solver) {}
+
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mlir::linalg::MatmulOp op,
+                                PatternRewriter &rewriter) const override {
+    // Determine if the left or right operand is secret to determine which
+    // matrix to diagonalize, or if both are secret or both are public, then
+    // return failure.
+    auto *leftOperandLookup =
+        solver->lookupState<SecretnessLattice>(op.getInputs()[0]);
+    Secretness leftOperandSecretness =
+        leftOperandLookup ? leftOperandLookup->getValue() : Secretness();
+    bool leftOperandIsSecret = (leftOperandSecretness.isInitialized() &&
+                                leftOperandSecretness.getSecretness());
+
+    auto *rightOperandLookup =
+        solver->lookupState<SecretnessLattice>(op.getInputs()[1]);
+    Secretness rightOperandSecretness =
+        rightOperandLookup ? rightOperandLookup->getValue() : Secretness();
+    bool rightOperandIsSecret = (rightOperandSecretness.isInitialized() &&
+                                 rightOperandSecretness.getSecretness());
+
+    LLVM_DEBUG({
+      llvm::dbgs() << "Left operand is secret: " << leftOperandIsSecret << "\n"
+                   << "Right operand is secret: " << rightOperandIsSecret
+                   << "\n";
+    });
+
+    // Error out if both are secret or both are public
+    if ((leftOperandIsSecret && rightOperandIsSecret) ||
+        (!leftOperandIsSecret && !rightOperandIsSecret)) {
+      return failure();
+    }
+    auto inputs = op.getInputs();
+    auto outputs = op.getOutputs();
+
+    Value secretValues;
+    Value publicMatrix;
+    if (leftOperandIsSecret) {
+      secretValues = inputs[0];
+      publicMatrix = inputs[1];
+    } else {  // the right operand is secret
+      secretValues = inputs[1];
+      publicMatrix = inputs[0];
+    }
```

Comment:
You could use a `std::swap`:
```suggestion
    // Assign if the left operand is secret
    Value secretValues = inputs[0];
    Value publicMatrix = inputs[1];
    if (rightOperandIsSecret) {
      std:swap(secretValues, publicMatrix);
    }
```